src/models/Album.js

```diff
@@ -1,61 +1,57 @@
 const mongoose = require('mongoose');
 
-const AlbumSchema = new mongoose.Schema({
-  artistName: {
-    type: String,
-    required:  [true, 'Please provide artist name'],
-    maxlength: [100, 'Artist name can not be more than 100 characters'],
-    trim: true,
-  },
-  albumName: {
-    type: String,
-    required:  [true, 'Please provide album name'],
-    maxlength: [100, 'Album name can not be more than 100 characters'],
-    trim: true,
-  },
-  price: {
-    type: Number,
-    required: [true, 'Please provide album price'],
-    min: 0,
-  },
-  image: {
-    type: String,
-    default: '/uploads/example.jpg',
-  },
-  releaseDate: {
-    type: Date,
-    required: false,
-  },
-  category: {
-    type: String,
-    required: [false, 'Please provide music category'],
-    enum: ['pop', 'rock', 'rap'],
-  },
-  spotifyUrl: {
-    type: String,
-    required: false, //CHANGE TO TRUE LATER!!!!!!
-    validate: {
-      validator: function (v) {
-        return /^https:\/\/open\.spotify\.com\/album\/[a-zA-Z0-9]+$/.test(v); //ensures that the URL starts with https://open.spotify.com/album/ followed by alphanumeric characters.
-      },
-      message: 'Invalid Spotify URL',
+const AlbumSchema = new mongoose.Schema(
+  {
+    artistName: {
+      type: String,
+      required: [true, 'Please provide artist name'],
+      maxlength: [300, 'Artist name can not be more than 100 characters'],
+      trim: true,
+    },
+    albumName: {
+      type: String,
+      required: [true, 'Please provide album name'],
+      maxlength: [300, 'Album name can not be more than 100 characters'],
+      trim: true,
+    },
+    price: {
+      type: Number,
+      required: [false, 'Please provide album price'],
+      default: 29,
+      min: 0,
+    },
+    image: {
+      type: String,
+      default: '/uploads/example.jpg',
+    },
+    releaseDate: {
+      type: Date,
+      required: false,
+    },
+    category: {
+      type: String,
+      required: [false, 'Please provide music category'],
+      enum: ['pop', 'rock', 'rap'],
+    },
+    spotifyUrl: {
+      type: String,
+      required: true,
+    },
+    averageRating: {
+      type: Number,
+      default: 0,
+    },
+    numOfReviews: {
+      type: Number,
+      default: 0,
+    },
+    user: {
+      type: mongoose.Types.ObjectId,
+      ref: 'User', // we are referencing the user model . if we don´t provide user- in postman we'll get  "msg": "Path `user` is required."
+      required: false, //CHANGE TO TRUE LATER!!!!!!
```

Comment:
In this code, we are saying that the album/product belongs to a single person. This _could_ be the case for something like seats in a concert venue, but in most online stores with physical products we likely want to be able to sell multiple copies of the same thing. So we end up with a "many-to-many" [relationship.](https://community.dbdiagram.io/t/tutorial-many-to-many-relationships/412) A user can buy multiple albums, and an album could be purchased by multiple users.

MongoDB is a No-SQL database so it doesn't have the usual way to do "joins" to handle this type of relationship. There's instead a few different ways we could do it: https://stackoverflow.com/a/46020968

1.  We could make this AlbumSchema have a `users` field, that is an array of `User` refs. That would allow us to do something like `album.users` to see all the users that bought a particular album. But then if we want to also be able to easily see all albums a particular user bought, then we need to also add to the `User` schema an `albums` field. And we need to remember to push to both of those whenever a purchase goes through. This is fine as long as we can make sure to remember to do so.
    1.  Here's a video tutorial that seems to do an okay job going over it: [https://www.youtube.com/watch?v=-El-86OzAhA](https://www.youtube.com/watch?v=-El-86OzAhA)
2.  We can create an intermediary schema (in a relational SQL table this would be called a "join table") like this:

```javascript
const PurchasedAlbumSchema = new Schema({
  albumId: { type: Schema.Types.ObjectId, ref: 'Album', required: true },
  userId: { type: Schema.Types.ObjectId, ref: 'User', required: true }
});
```

Then when someone buys an album, you'd record it like:

```javascript
await PurchasedAlbum.create({ albumId: <id>, userId: <id> });
```

For option number 2 there's two ways to handle fetching that info:

1) Virtuals:

```javascript
AlbumSchema.virtual('purchasedByUsers', {
  ref: 'PurchasedAlbum',
  localField: '_id',
  foreignField: 'userId'
});

UserSchema.virtual('purchasedAlbums', {
  ref: 'PurchasedAlbum',
  localField: '_id',
  foreignField: 'albumId'
});

# then in the code can use populate()
// Show user by id with all the albums they've purhcased
    let user = await User.findOne("id").populate({
      path: 'purchasedAlbums',
      populate: { path: 'albumId' }
    });
```

2) Aggregate queries:

```javascript
const userId = 'your_user_id';

try {
 const result = await PurchasedAlbum.aggregate([
  {
    $match: { userId: mongoose.Types.ObjectId(userId) }
  },
  {
    $lookup: {
      from: 'Album',
      localField: 'albumId',
      foreignField: '_id',
      as: 'albums'
    }
  },
  {
    $unwind: '$albums'
  },
  {
    $project: {
      _id: 0,
      userId: 1,
      albums: 1
    }
  }
]).exec();

console.log(result);

} catch (err) {
  logger.error(err);
}
```

Reply:
@akosasante thank you! i've just tried to add virtuals- could you please take a look https://github.com/Code-the-Dream-School/dd-prac-team4-back/pull/15/commits/45e24d40b1ccf778befa9c8cc609ab0f4d4751be
should we add await PurchasedAlbum.create({ albumId: <id>, userId: <id> }); to ordersControllers.js or maybe i am missing smth and should add it somewhere else?

Reply:
1.  PurchasedAlbum model:

Apart from my other earlier comments, in the example below I've also renamed the properties from userId/albumId to user/album, just to make it clearer when we do the `populate` call that we'll usually be working with these as the full object rather than jsut the id (except maybe when inserting)

```javascript
const mongoose = require('mongoose');

const PurchasedAlbumSchema = new mongoose.Schema({
  album: { type: mongoose.Schema.Types.ObjectId, ref: 'Album', required: true },
  user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
});

const PurchasedAlbum = mongoose.model('PurchasedAlbum', PurchasedAlbumSchema);

module.exports = PurchasedAlbum;
```

2\. Virtual in the User model:

I would rename the virtual to be a bit clearer, that this is returning _all_ the albums this single _user_ has purchased

```javascript
UserSchema.virtual('purchasedAlbums', {
  ref: 'PurchasedAlbum',
  localField: '_id',
  foreignField: 'user'
});
```

3\. Virtual in the Album model

I would rename the virtual to be a bit clearer, that this is returning all of the users that have purchased _this album_

```
AlbumSchema.virtual('purchasedByUsers', {
  ref: 'PurchasedAlbum', //specifies that the virtual field purchasedAlbums is referencing the PurchasedAlbum model.
  localField: '_id', //_id field (from the db= id of the product) of the Album model is used as the local field to establish the relationship.
  foreignField: 'album' // album field in the PurchasedAlbum model is used as the foreign field to establish the relationship.
});
```

Now we can do all sorts of queries, like:

1.  Fetching a user from the database, including all the albums they've purchased
2.  Fetching an album from the database, including all the users that have purchased it
3.  Fetching a list of purchased albums for a user
4.  Fetching a list of users who purchased an album
5.  Saving to the database that a user purchased a particular album

```javascript
// 1 - example A:
const getCurrentUserWithPurchasedAlbums = async (req, res) => {
  // Show current user by id with all the albums they've purchased
  let usersWithAlbums = await User.findById(req.user.userId).populate({
      path: 'purchasedAlbums', // name of the virtual to populate
      populate: { path: 'album' } // nested populate, without this we would just get back a list of PurchasedAlbum models. 
      // But we just want to further populate to get the Album model refferred to in  the PurchasedAlbum.album proprty.
    });
    res.status(StatusCodes.OK).json({ usersWithAlbums, count: usersWithAlbums.length });
  }

// 1 - example B (updating existing getUser endpoint to take query parameters)
const getSingleUser = async (req, res) => {
  // Function to get a single user by ID
  // Find the user in the database based on the provided user ID and exclude the 'password' field
  let query = User.findOne({ _id: req.params.id }).select('-password');
  if (req.query.includeAlbums === "true") {
    query = query.populate({
      path: 'purchasedAlbums', // name of the virtual to populate
      populate: { path: 'album' } // nested populate, without this we would just get back a list of PurchasedAlbum models. 
      // But we just want to further populate to get the Album model refferred to in  the PurchasedAlbum.album proprty.
    });
  }
  const user = await query;
  if (!user) {
    // Throw a NotFoundError if the user is not found
    throw new CustomError.NotFoundError(`No user with id: ${req.params.id}`);
  }
  checkPermissions(req.user, user._id); // Check if the user has permission to access the user's information
  res.status(StatusCodes.OK).json({ user }); // Send a JSON response with the status code 200 OK and the user
};

// 2 - Same as above but you would populate with the object { path: 'purchasedByUsers', populate: { path: 'user' }}

// 3 - await PurchasedAlbum.find({user: user._id}).populate('album')
Returns a list of PurchasedAlbum records, with the album populated with the full model instance:
[
  {
    _id: new ObjectId("64d91dbece2137867e2f40d3"),
    album: {
      _id: new ObjectId("64d2a94c793389a43fc5a8d5"),
      artistName: 'Travis Scott',
      albumName: 'UTOPIA',
      price: 19.99,
      image: 'https://i.scdn.co/image/ab67616d0000b273881d8d8378cd01099babcd44',
      releaseDate: 2023-07-28T00:00:00.000Z,
      spotifyUrl: 'https://api.spotify.com/v1/albums/18NOKLkZETa4sWwLMIm0UZ',
      averageRating: 0,
      numOfReviews: 0,
      __v: 0,
      createdAt: 2023-08-08T20:45:00.942Z,
      updatedAt: 2023-08-13T17:56:20.086Z
    },
    user: new ObjectId("64d44bc8337399ccf9ad7e4d"),
    __v: 0
  }
]

// 4 - same idea, but instead `await PurchasedAlbum.find({album: albumId}).populate('user')

// 5 -  await PurchasedAlbum.create({albumId: album._id, userId: user._id})

```

And yes for (5) we would probably end up putting that somewhere in an `OrdersController`. We would save the order model. Then once the user's payment has gone through successfully and the order is permanent, we can do the code in (5) to store their purchased albums. 

(1) and (3) will probably be in the `UsersController`. (2) and (4) in the `AlbumsController`.